src/nspawn/nspawn-mount.c

```diff
@@ -628,11 +679,105 @@ static int mount_legacy_cgroup_hierarchy(const char *dest, const char *controlle
         return 1;
 }
 
-static int mount_legacy_cgroups(
+// Mount a legacy cgroup hierarchy when cgroup namespaces are supported.
+static int mount_legacy_cgns_supported(
+		bool userns, uid_t uid_shift, uid_t uid_range,
+		const char *selinux_apifs_context) {
+	_cleanup_set_free_free_ Set *controllers = NULL;
+	const char *cgroup_root = "/sys/fs/cgroup";
+	int r;
+
+	(void) mkdir_p(cgroup_root, 0755);
+
+	/* Mount a tmpfs to /sys/fs/cgroup if it's not mounted there yet. */
+	r = path_is_mount_point(cgroup_root, AT_SYMLINK_FOLLOW);
+	if (r < 0)
+		return log_error_errno(r, "Failed to determine if /sys/fs/cgroup is already mounted: %m");
+	if (r == 0) {
+		_cleanup_free_ char *options = NULL;
+
+		/* When cgroup namespaces are enabled and user namespaces are used then the mount of the cgroupfs is
+		 * done *inside* the new user namespace. We're root in the new user namespace and the kernel will
+		 * happily translate our uid/gid to the correct uid/gid as seen from e.g. /proc/1/mountinfo. So we
+		 * simply pass uid 0 and not uid_shift to tmpfs_patch_options().
+		 */
+		r = tmpfs_patch_options("mode=755", userns, 0, uid_range, selinux_apifs_context, &options);
+		if (r < 0)
+			return log_oom();
+
+		if (mount("tmpfs", cgroup_root, "tmpfs", MS_NOSUID|MS_NOEXEC|MS_NODEV|MS_STRICTATIME, options) < 0)
+			return log_error_errno(errno, "Failed to mount /sys/fs/cgroup: %m");
+	}
+
+	if (cg_unified() > 0)
+		goto skip_controllers;
+
+	controllers = set_new(&string_hash_ops);
+	if (!controllers)
+		return log_oom();
+
+	r = get_controllers(controllers);
+	if (r < 0)
+		return log_error_errno(r, "Failed to determine cgroup controllers: %m");
+
+	for (;;) {
+		char *saveptr = NULL, *tok = NULL;
+		_cleanup_free_ char *controller = NULL, *origin = NULL, *target = NULL;
+
+		controller = set_steal_first(controllers);
+		if (!controller)
+			break;
+
+		if (userns)
+			r = mount_legacy_cgroup_hierarchy("", controller, controller, false);
+		else
+			r = mount_legacy_cgroup_hierarchy("", controller, controller, true);
+		if (r < 0)
+			return r;
+
+		origin = strdup(controller);
+		if (!origin)
+			return log_oom();
+
+		// When multiple hierarchies are co-mounted, make their constituting individual hierarchies a symlink
+		// to the co-mount.
+		for (tok = strtok_r(controller, ",", &saveptr); tok;
+				tok = strtok_r(NULL, ",", &saveptr)) {
```

Comment:
hmpf, any chance you can replace strok_r usage by a loop around extract_first_word()?